pkg/kubectl/cmd/create_deployment.go

```diff
@@ -76,25 +135,21 @@ func CreateDeployment(f cmdutil.Factory, cmdOut, cmdErr io.Writer, cmd *cobra.Co
 	if err != nil {
 		return fmt.Errorf("failed to discover supported resources: %v", err)
 	}
+
 	generatorName := cmdutil.GetFlagString(cmd, "generator")
-	// fallback to the old generator if server does not support apps/v1beta1 deployments
-	if generatorName == cmdutil.DeploymentBasicAppsV1Beta1GeneratorName &&
-		!contains(resourcesList, appsv1beta1.SchemeGroupVersion.WithResource("deployments")) {
-		fmt.Fprintf(cmdErr, "WARNING: New deployments generator specified (%s), but apps/v1beta1.Deployments are not available, falling back to the old one (%s).\n",
-			cmdutil.DeploymentBasicAppsV1Beta1GeneratorName, cmdutil.DeploymentBasicV1Beta1GeneratorName)
-		generatorName = cmdutil.DeploymentBasicV1Beta1GeneratorName
-	}
-	var generator kubectl.StructuredGenerator
-	switch generatorName {
-	case cmdutil.DeploymentBasicAppsV1Beta1GeneratorName:
-		generator = &kubectl.DeploymentBasicAppsGeneratorV1{Name: name, Images: cmdutil.GetFlagStringSlice(cmd, "image")}
-	case cmdutil.DeploymentBasicV1Beta1GeneratorName:
-		generator = &kubectl.DeploymentBasicGeneratorV1{Name: name, Images: cmdutil.GetFlagStringSlice(cmd, "image")}
-	default:
-		return cmdutil.UsageError(cmd, fmt.Sprintf("Generator: %s not supported.", generatorName))
+
+	// It is possible we have to modify the user-provided generator name if
+	// the server does not have support for the requested generator.
+	generatorName = fallbackGeneratorNameIfNecessary(generatorName, resourcesList, cmdErr)
+
+	imageNames := cmdutil.GetFlagStringSlice(cmd, "image")
+	generator, ok := generatorFromName(generatorName, imageNames, deploymentName)
```

Comment:
why return true/false here instead of error, also if there's only one invoke here, I don't think it's needed to abstract those logic for function `generatorFromName` here.